Feat/custom moves

cptleo92 opened the pull request:
#60 

GMs can add/edit/remove custom moves for a campaign. For now, I figured that the section can live in the campaign screen so that everyone can view what's available.

At this point, I'm not too sure how to handle showing these custom moves in the character move list. My initial plan was to somehow add in all these custom moves in `useFilterMoves()`. I'm pretty sure these "custom moves" need to be in a store of some kind, maybe as an addition to the campaign store. 

Let me know what your thoughts are on moving forward with this approach!

cptleo92 committed:
Added firestore functionality for custom campaign moves

cptleo92 committed:
Created campaign custom move components

cptleo92 committed:
Added delete custom move from campaign button

cptleo92 committed:
Refactored CustomMoveDialog to allow editing/updating

scottbenton commented:
Looking through the code, it looks like you've done a great job as always matching the existing structure of the API calls. Nice work! 

Looks like we run into an error calling `updateDoc` if the document doesn't exist yet - you might need to create it first. I do something similar [here](https://github.com/scottbenton/Iron-Fellowship/blob/dev/src/api/user/settings/listenToOracleSettings.ts#L19) - if I don't get a document back from firestore in the listener, I create one.

I agree with your approach moving forward, attaching them to the campaign store, and then appending them to the existing moves in useFilterMoves() store makes sense - probably adding a specific "Custom Moves" section to the moves list.

The only other note I have is that I would prefer for this section to be added to the CampaignGMScreenPage, but I can always move it later if that is too much of a pain.

cptleo92 committed:
Changed updateDoc to setDoc in case of no document found

cptleo92 committed:
Moved custom moves section to GM page

cptleo92 commented:
This takes care of moving the section to CampaignGMScreenPage and the updateDoc issue.

Looking at how everything works, it seems like anytime a user's campaign is updated, the store is updated at the same time -- based on my understanding of how `listenToUsersCampaigns.ts` works. 

In Firestore right now, moves are stored under a Settings collection, which is nested in the Campaign collection. I think this makes things a bit more complicated because I'd have to set up a store, update it on doc updates, etc. 

What if instead, settings was a field directly in the campaign document? Something like this?

```js
{
    "gmId": "Hn7mcJWtwqYB1mI9hYIPr6vr2K62",
    "characters": [
        {
            "uid": "Hn7mcJWtwqYB1mI9hYIPr6vr2K62",
            "characterId": "XKM6qTY2okhZMjsWCLx0"
        }
    ],
   ...
   ...
   "settings": {
       "moves": []
   }
}
```

I haven't tested it out either way, but it does seem like this will simplify things a bit. Of course, if you prefer the database a certain way, I'm happy to implement it any way. Just wanted to pick your brain on this!

scottbenton commented:
Hey sorry for taking so long to get back to you on this! I like the way you have the database set up right now. The lighter the campaign document is, the quicker the initial app load will be. Its not going to make a big difference (or even a noticeable one haha), but I think keeping them separate lowers the amount of clutter we load in.

From here, it looks like you just need to add them to the moves lists on the campaign and character screens. Nice work!

scottbenton commented:
And of course, let me know if you have any more questions :)

cptleo92 commented:
No worries, I'm not in any rush on this! I'll finish this up. 

In the meantime, I was curious about your `useApiState()` hook. What's your reason for creating it? It seems like it helps with keeping things more organized, but is there a performance benefit as well?

cptleo92 committed:
Added settings store, minor tweaks on naming

scottbenton commented:
No performance benefit at all - its more to have error handling and loading states handled the same on all api calls. I realized I was rewriting a lot of the same code everywhere, so I tried to condense it. Although now I'm rewrite the `useAddXYZ` functions and such everywhere, so... at least its more organized than it was before?

scottbenton commented:
If you have any ideas on how to improve this, or any of the other code you see throughout the app as you are working, please let me know! I'm more happy with this method than I was before, but it is by no means perfect.

cptleo92 commented:
I have no recommendations! I was just curious to know what motivated it. It does remind me a bit of [React Query](https://react-query-v3.tanstack.com/) since your hook and the library handles data, loading, and errors similarly.

cptleo92 committed:
Refactored useFilterMoves() to pull custom moves from store

cptleo92 committed:
Add customMoves to useFilterMoves dependency to re-render moves list on updates

cptleo92 committed:
Remove unused useEffect from MovesSection

cptleo92 commented:
I created a settings store and refactored `useFilterMoves()` to take in a `campaignId`, which it uses to pull custom moves information. I added `customMoves` to the `useFilterMoves()` dependency array so that the moves list stays up to date with the custom moves, though I'm not 100% sure that this is an optimal way of handling that. 

I did consider what happens when a campaign gets deleted. I figured that since `settings/moves` is a subcollection of `campaigns`, if a campaign is removed then so will the settings naturally. That being the case, I didn't implement any deletion functions for settings. 

Let me know your thoughts and any improvements I can make!

scottbenton committed:
Merge branch 'dev' into feat/custom-moves

scottbenton reviewed (approved):
Hey thank you for your work on this! This is great.

There are a couple of things I'm writing comments on in the code just so you get some feedback if you want it, but I'm going to merge anyway and fix them up on my side, because my next changes are going to cause some chaos and I don't want to make you have to merge this into the middle of that haha.

Thanks as always for your help! My friend has been asking for custom moves for a bit, so this is super helpful.

scottbenton commented on src/components/MovesSection/MovesSection.tsx:
```diff
@@ -6,21 +6,23 @@ import useFilterMoves from "./useFilterMoves";
 export function MovesSection(props: MovesSectionProps) {
-  const { stats } = props;
+  const { stats, campaignId } = props;
+  if (campaignId) useListenToCampaignCustomMoves(campaignId);
```
Hooks should not be called conditionally - instead, useListenToCampaignCustomMoves should allow campaignId to be undefined.

This is because react relies on hooks being called in the same order on every single rerender. You can read more about that [here in the react docs](https://reactjs.org/docs/hooks-rules.html#explanation)

cptleo92 replied:
I definitely need to read through the docs more carefully. Knowing just this one thing has cleared up a lot of confusion for issues I've had in the past. Thanks!

cptleo92 commented:
Awesome, glad it worked out! This was quite the challenge for me but I learned a lot so thanks for letting me give this a shot.
